pmd-core/src/main/java/net/sourceforge/pmd/util/FileFinder.java

```diff
@@ -17,10 +19,25 @@
     private FilenameFilter filter;
     private static final String FILE_SEP = System.getProperty("file.separator");
 
+    /**
+     * Searches for files in a given directory.
+     * The returned files are sorted alphabetically by path, ignoring the case.
+     *
+     * @param dir     the directory to search files
+     * @param filter  the filename filter that can optionally be passed to get files that match this filter
+     * @param recurse search for files recursively or not
+     * @return list of the found files sorted alphabetically by path, ignoring the case
+     */
     public List<File> findFilesFrom(File dir, FilenameFilter filter, boolean recurse) {
         this.filter = filter;
         List<File> files = new ArrayList<>();
         scanDirectory(dir, files, recurse);
+        Collections.sort(files, new Comparator<File>() {
+            @Override
+            public int compare(File o1, File o2) {
+                return o1.getPath().compareToIgnoreCase(o2.getPath());
```

Comment:
While digging further into this sorting, I found out that `java.io.File` implements Comparable and would sort Files correctly depending on the operating system.

So, we could even just say `Collections.sort(files);`.

However, we can also change the method `scanDirectory` below instead to do the sorting already:

```java
File[] candidates = dir.listFiles(filter);
if (candidates == null) {
    return;
}

Arrays.sort(candidates);
...
```

Using File instead of String here has the advantage, that we don't need to create new File instances on our own.

Reply:
> to analyse the files always on the same order to get always the same output, independently of
> the type of filesystem where CPD is executing.

That's an interesting requirement. I first concentrated on making sure, repeatable runs would produce the same result (so we just need some sorting).

I just tried it out: File.compareTo will sort this way:
On Windows: `zzzz3.txt`, then `zZzz4.txt`
On Linux: `zZzz4`, then `zzzz3`

Which would indeed result in different output...

So, you're right and we'll need to use [PATH_INSENSITIVE_COMPARATOR](https://commons.apache.org/proper/commons-io/javadocs/api-release/org/apache/commons/io/comparator/PathFileComparator.html#PATH_INSENSITIVE_COMPARATOR)

Reply:
Thanks for checking it! I updated the PR to use the PATH_INSENSITIVE_COMPARATOR and moved the sort to the `scanDirectory` method like you suggested. 🙂

Reply:
Awesome, thanks!
I'll add a comment so that we remember in the future, why we want a case insensitive sorting here.